In [35]:
import numpy as np
import matplotlib.pyplot as plt

## Opgave 11.1

$$
\begin{bmatrix}
x^{3} & x^{2} & x & 1
\end{bmatrix}

\begin{bmatrix}
x_{0} \\
x_{1} \\
x_{2} \\
x_{3}
\end{bmatrix}

=

\begin{bmatrix}
(x - 1)^{3}
\end{bmatrix}
$$
$$
\begin{bmatrix}
x^{3} & x^{2} & x & 1
\end{bmatrix}

\begin{bmatrix}
x_{0} \\
x_{1} \\
x_{2} \\
x_{3}
\end{bmatrix}

=

\begin{bmatrix}
x^{3} & - 3 x^{2} & 3 x & - 1 
\end{bmatrix}
$$

$$
\begin{bmatrix}
x_{0} \\
x_{1} \\
x_{2} \\
x_{3}
\end{bmatrix}

=

\begin{bmatrix}
1 \\
-3 \\
3 \\
-1
\end{bmatrix}
$$

## Opgave 11.2

### (a)

In [36]:
def back_subs(u, b):
    _, n = u.shape
    x = np.empty((n, 1))
    for i in reversed(range(n)):
        x[i] = (b[i] - u[[i], i+1:] @ x[i+1:]) / u[i, i]
    return x

In [37]:
p = np.array([1, -1, 2, 4])[:, np.newaxis]

E = np.array([  [1, 1, 1, 1],
                [0, 1, 1, 1],
                [0, 0, 1, 1],
                [0, 0, 0, 1]])

back_subs(E, p)

array([[ 2.],
       [-3.],
       [-2.],
       [ 4.]])

### (b)

In [38]:
F = np.array([  [1/2,          1/np.sqrt(2), 0,            1/2],
                [-1/2,         1/np.sqrt(2), 0,           -1/2],
                [1/2,          0,            1/np.sqrt(2),-1/2],
                [-1/2,         0,            1/np.sqrt(2), 1/2]])

np.linalg.solve(F, p)

array([[0.        ],
       [0.        ],
       [4.24264069],
       [2.        ]])

In [39]:
# transponeret er det samme som inverse, fordi F matricen er ortogonal
F.T @ p

array([[0.        ],
       [0.        ],
       [4.24264069],
       [2.        ]])

In [40]:
S = np.linalg.inv(F) @ E
S

array([[ 0.5       ,  0.        ,  0.5       ,  0.        ],
       [ 0.70710678,  1.41421356,  1.41421356,  1.41421356],
       [ 0.        ,  0.        ,  0.70710678,  1.41421356],
       [ 0.5       ,  0.        , -0.5       ,  0.        ]])

## Opgave 11.3

### (a)

In [41]:
A = np.array([  [3, 6, -1, 3, 1 ],
                [1, 2, 1, -1, -1],
                [2, 4, 0,  1, 1 ]], dtype=float)

v0 = np.copy(A[:,0,np.newaxis])
v1 = np.copy(A[:,2,np.newaxis])
v2 = np.copy(A[:,4,np.newaxis])

A[0] -= A[2]
A[1] -= A[0]
A[1] *= 1/A[1,2]
A[2] -= 2 * A[0]
A[2] -= 2 * A[1]
A[2] *= 1/A[2,4]
A[0] += A[1]
A[0] += A[2]/2
A[1] += A[2]/2
A

array([[ 1. ,  2. ,  0. ,  0.5,  0. ],
       [ 0. ,  0. ,  1. , -1.5,  0. ],
       [ 0. ,  0. ,  0. ,  0. ,  1. ]])

In [42]:
print(f"Søjlerummet S(A) kommer til at være udspændt af følgende vektorer:\n{v0}\n\n{v1}\n\n{v2}")

Søjlerummet S(A) kommer til at være udspændt af følgende vektorer:
[[3.]
 [1.]
 [2.]]

[[-1.]
 [ 1.]
 [ 0.]]

[[ 1.]
 [-1.]
 [ 1.]]


$$
S(A) = \text{span}\{  
\begin{bmatrix}
3 \\
1 \\
2
\end{bmatrix}
,
\begin{bmatrix}
-1 \\
1 \\
0
\end{bmatrix}
,
\begin{bmatrix}
1 \\
-1 \\
1
\end{bmatrix}
\}
$$

Følgende ligninger opstilles:
$$
\begin{align*}
x_{0} + 2 x_{1} + 0.5 x_{3} = 0 & \Leftrightarrow x_{0} = -2 x_{1} - 0.5 x_{3} \\
x_{2} - 1.5 x_{3} = 0 & \Leftrightarrow x_{2} = 1.5 x_{3} \\
x_{4} & = 0
\end{align*}
$$

Den generelle løsning bliver:
$$
\begin{bmatrix}
x_{0} \\
x_{1} \\
x_{2} \\
x_{3} \\
x_{4}
\end{bmatrix}
=
\begin{bmatrix}
-2 x_{1} - 0.5 x_{3} \\
x_{1} \\
1.5 x_{3} \\
x_{3} \\
0
\end{bmatrix}
=
x_{1} 
\begin{bmatrix}
-2 \\
1 \\
0 \\
0 \\
0
\end{bmatrix}
+
x_{3}
\begin{bmatrix}
-0.5 \\
0 \\
1.5 \\
1 \\
0
\end{bmatrix}
$$

Nulrummet N(A) er da givet ved udspændingen af vektorerne således:
$$
N(A) = \text{span} \{
\begin{bmatrix}
-2 \\
1 \\
0 \\
0 \\
0
\end{bmatrix}
,
\begin{bmatrix}
-0.5 \\
0 \\
1.5 \\
1 \\
0
\end{bmatrix}
\}
$$

### (b)

Vektoren kan opstilles ud fra baserne således:
$$
0 \cdot
\begin{bmatrix}
3 \\
1 \\
2
\end{bmatrix}
+
1 \cdot 
\begin{bmatrix}
-1 \\
1 \\
0
\end{bmatrix}
+
1 \cdot
\begin{bmatrix}
1 \\
-1 \\
1
\end{bmatrix}
=
\begin{bmatrix}
0 \\
0 \\
1
\end{bmatrix}
$$

Den må da være en del af søjlerummet. 

### (c)

$$
x
=
\begin{bmatrix}
0 \\
0 \\
1 \\
0 \\
1
\end{bmatrix}
+
c_{0}
\begin{bmatrix}
-2 \\
1 \\
0 \\
0 \\
0
\end{bmatrix}
+
c_{1}
\begin{bmatrix}
-0.5 \\
0 \\
1.5 \\
1 \\
0
\end{bmatrix}
$$